pkg/ebpf/tcp_dns_layer.go

```diff
@@ -0,0 +1,52 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"encoding/binary"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+// source: https://github.com/weaveworks/scope/blob/master/probe/endpoint/dns_snooper.go
+// Gopacket doesn't provide direct support for DNS over TCP, see https://github.com/google/gopacket/issues/236
+type tcpWithDNSSupport struct {
+	tcp layers.TCP
+}
+
+func (m *tcpWithDNSSupport) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	return m.tcp.DecodeFromBytes(data, df)
+}
+
+func (m *tcpWithDNSSupport) CanDecode() gopacket.LayerClass { return m.tcp.CanDecode() }
+
+// Determine if a TCP segment contains a full DNS message (i.e. not fragmented)
```

Comment:
I think you could improve the comments here and explain the "heuristic" we're using here.
From what I understand, when you have a TCP segment enclosing a DNS payload you prepend it with 2 bytes indicating the size of the DNS payload, and we consider it to be "self-contained" when the length of payload  matches that.